docs/source/guides/mod-config.md

# Mod Configuration Menu

MWSE comes with a built-in framework for easily making a configuration menu for your mod. All configuration menus made in this way are accessible by pressing the "Mod Config" button in the pause menu. This guide covers the basic concepts and syntax of the Mod Configuration Menu (MCM), and provides some examples to help you get started. More in-depth information can be found on the [`mwse.mcm` API page](../apis/mwse.mcm.md).

!!! note
	The `mwse.mcm` API acts as a complete replacement for the [easyMCM](https://easymcm.readthedocs.io/en/latest/) framework. It is backwards compatible with easyMCM, and provides several new functionalities.
	It's recommended that you use `mwse.mcm` instead of easyMCM.

## Simplest Example
Here is the simplest code required to create a functional mod config:

```lua linenums="1"
-- Default settings for your mod.
local myDefaultConfig = {
	enabled = true
}

-- The name of the config json file of your mod.
-- This is used to save/load your config, so that changes persist 
-- between game launches.
local configPath = "My Awesome Mod" 

-- Loaded config, taking into account any setting changes made by the user.
local myConfig = mwse.loadConfig(configPath, defaultConfig)

-- When the mod config menu is ready to start accepting registrations,
-- register this mod.
local function registerModConfig()
	-- Create the top level component Template. 
	-- This is basically a box that holds all the other parts of the menu.
	local template = mwse.mcm.createTemplate({ 
		-- This will be displayed in the mod list on the lefthand pane.
		name = "My Awesome Mod", 
		-- This makes all settings update the values stored in this table.
		config = myConfig 
	})

	-- This tells MWSE to add this config menu to the list.
	template:register()

	-- Saves the config to a file whenever the menu is closed.
	template:saveOnClose(configPath, myConfig)

	-- Create a simple container Page under Template.
	local page = template:createPage({ label = "Settings" })

	-- Create a button under that controls the `"enabled"` 
	-- setting in `myConfig`.
	-- Clicking on this button will change `"enabled"` from `true`
	-- to `false, and vic-versa.
	page:createYesNoButton({ 
		-- The `label` is the text that will be shown in your config menu.
		-- It lets users know what the button does.
		label = "Enable Mod", 
		configKey = "enabled" 
	})
end
event.register(tes3.event.modConfigReady, registerModConfig)
```

After loading up the game, you should see "My Awesome Mod" in the mod config menu, with a Yes/No button with label "Enable Mod" in its configuration pane.


## Other Sorts of Settings
In practice, most mods need more than simple "yes" and "no" buttons. Fortunately, MWSE includes many different sorts of settings that help to cover pretty much any use-case.

Here's an example that comes up fairly frequently: some part of your mod involves making a new UI element, and you'd like to let the user control where things show up on the screen.
The [`PercentageSlider`](../types/mwseMCMPercentageSlider.md) type is well suited to this task. Let's see how we can use it to extend the previous example.

```lua hl_lines="4 5 6 28 29 30 31 32 33 34 35 36 37 38 39 40"
-- Default settings for your mod.
local myDefaultConfig = {
	enabled = true,
	-- These values should be between 0 and 1.
	xPosition = 0.5, -- 0 means "left-most", and 1 means "right-most".
	yPosition = 0.5, -- 0 means "top-most", and 1 means "bottom-most".
}

local configPath = "My Awesome Mod" 
local myConfig = mwse.loadConfig(configPath, defaultConfig)

-- When the mod config menu is ready to start accepting registrations,
-- register this mod.
local function registerModConfig()
	
	local template = mwse.mcm.createTemplate({ 
		name = "My Awesome Mod", 
		config = myConfig
	})
	template:register()
	template:saveOnClose(configPath, myConfig)

	-- Create a simple container Page under Template.
	local page = template:createPage({ label = "Settings" })

	page:createYesNoButton({ label = "Enable Mod", configKey = "enabled" })

	-- Create some sliders that let the user control the positioning of whatever menu element
	-- your mod is making.
	-- Notice how the syntax is exactly the same as the syntax used to create new buttons.
	-- The only difference is we're now telling the page we want a `PercentageSlider`
	-- instead of a `YesNoButton`.
	page:createPercentageSlider({ 
		label = "Menu X Position", 
		configKey = "xPosition"
	})
	page:createPercentageSlider({ 
		label = "Menu Y Position", 
		configKey = "yPosition"
	})
end
event.register(tes3.event.modConfigReady, registerModConfig)
```
You should now see two additional sliders added to your mod.

It's also pretty common that mods would like certain things to happen whenever a key is pressed. 
So it would be nice if the MCM had some easy way of managing keybinds... Luckily, it does!

```lua hl_lines="8 9 10 11 12 13 14 15 46 47"
-- Default settings for your mod.
local myDefaultConfig = {
	enabled = true,
	-- These values should be between 0 and 1.
	xPosition = 0.5, -- 0 means "left-most", and 1 means "right-most".
	yPosition = 0.5, -- 0 means "top-most", and 1 means "bottom-most".

	-- This stores the key combination that opens your mods UI element.
	-- This value corresponds to holding `Alt + H`.
	menuKeyCombo = {
		keyCode = tes3.scanCode.h,
		isAltDown = true,
		isControlDown = false,
		isShiftDown = false
	},
}

local configPath = "My Awesome Mod" 
local myConfig = mwse.loadConfig(configPath, defaultConfig)

-- When the mod config menu is ready to start accepting registrations,
-- register this mod.
local function registerModConfig()
	-- Create the top level component Template. This is basically a box that holds all the other parts of the menu.
	
	local template = mwse.mcm.createTemplate({ name = "My Awesome Mod", config = myConfig })
	template:register()
	template:saveOnClose(configPath, myConfig)

	-- Create a simple container Page under Template.
	local page = template:createPage({ label = "Settings" })

	-- Create a button under Page that toggles a variable between true and false
	page:createYesNoButton({ label = "Enable Mod", configKey = "enabled" })

	-- Controls the position of the menu.
	page:createPercentageSlider({ 
		label = "Menu X Position", 
		configKey = "xPosition"
	})
	page:createPercentageSlider({ 
		label = "Menu Y Position", 
		configKey = "yPosition"
	})

	-- Controls the key that opens the menu.
	page:createKeyBinder({ label = "Menu key", configKey = "menuKeyCombo" })
end
event.register(tes3.event.modConfigReady, registerModConfig)
```


## Understanding the Terminology
Now that we've covered a few examples, it might be helpful to explain the big picture a bit.

The MCM is made up of things called **components**. In the most simple setup, the MCM consists of three different types of components:

1. Various **Settings**. These are the reason you wanted to make a menu in the first place. 
	- They let users change various parts of a `config` table by interacting with various UI elements.
	- There are many different types of settings: you can have dropdown menus, sliders, buttons, keybinders, and more.
2. A few **Pages**. These hold various settings, and let you group them into separate pages.
	- In the simple examples we covered above, we only had at most four settings, so there wasn't really a need for multiple pages.
		- But as your mod grows, you might find yourself adding a lot of settings to the MCM. 
		- In those situations, it can be nice to split things up across several pages.
	- There are a few different types of `Pages`, each suited to different use-cases.
3. One **Template**. This is the thing that holds all the various pages of your mod, as well as some important information about your mod.
	- It needs to know the `name` of your mod, so it can be properly placed in the list on the left side of the menu.
	- It can store your mods `config` table, and can facilitate the process of saving changes to a file whenever the menu closes.

All MCMs will consist of those three types of components. They're building blocks of any mod's config menu. 
But there are a couple other components that can be useful:

4. **Categories**. These basically work like subsections of a `Page`. 
	- You can add a `Category` to a page as a way to bundle similar settings together.
	- You can even make categories inside of other categories.
5. **Infos**. These can sometimes show up in specific situations as a way to present information that's not directly tied to a specific config setting.

## "Advanced" Examples




### Side Bar Page

MCM has a lot of features, and one of the most commonly used components is the `SideBarPage`. When a component is moused over in a `SideBarPage`, the right panel will display the `description` of that component. It is recommended to use a `SideBarPage` instead of a regular `Page` whenever a mod has a lot of settings.

```lua title="My Awesome Mod/main.lua"
	local settings = template:createSideBarPage({
		label = "Settings",
		-- This description will be displayed at the right panel if no component is moused over.
		-- (The `\z` character lets you define a string across multiple lines.)
		description =
			"My Awesome Mod\n\z
			Version 1.0.0\n\z
			\n\z
			The mod is very awesome. And this is its awesome description.\n",
	})

	settings:createYesNoButton({
		label = "Enable Mod",
		-- This description will be displayed at the right panel if the button is moused over.
		description =
			"If this setting is enabled, the mod is enabled.\n\z
			\n\z
			If this setting is disabled, the mod is disabled."
		variable = mwse.mcm:createTableVariable({ id = "enabled", table = config }),
	})
```

### Displaying Default Values of Settings
As your config menu grows, you may want to include some information about what the default values for various settings are. 
Fortunately, the MCM comes with a very straight-forward solution to this problem:
```lua hl_lines="4 5"
local template = mwse.mcm.createTemplate({ 
	name = "My Awesome Mod", 
	config = myConfig,
	defaultConfig = myDefaultConfig,
	showDefaultSetting = true,
})
```

This will result in the default values of settings automatically being displayed underneath their descriptions, provided those settings are in a `SideBarPage`.

### Separate config.lua

As your default config gets long, you might want to create a separate Lua file for it.

```lua linenums="1" title="My Awesome Mod/config.lua"
local defaultConfig = {
	enabled = true,
	keybind = {
		keyCode = tes3.scanCode.k,
		isShiftDown = false,
		isAltDown = false,
		isControlDown = true,
	}, -- Ctrl + K
	nonEdible = {
		["ingred_diamond_01"] = true,
		["ingred_scrap_metal_01"] = true,
	}
}
local config = mwse.loadConfig("My Awesome Mod", defaultConfig)
-- Return both the `config` and the `defaultConfig`, so that the MCM can properly
-- display default values.
return config, defaultConfig
```

We can then use the `require` function to import the `config` and `defaultConfig` in `main.lua`.

```lua linenums="1" title="My Awesome Mod/main.lua"
local config, defaultConfig = require("MyMod.config")
```

!!! note
	Having a dedicated config file is important when your mod has multiple files.
	When using a dedicated config file, you can import it from every file of your mod using `require`, and all of those files
	will use a sychronized copy of your config.

### Exclusions Page

Let's say your Awesome Mod makes certain ingredients non-edible. Using an Exclusions Page, you can allow users to manually add what ingredients should be non-edible in game.

```lua linenums="1" title="My Awesome Mod/main.lua"
template:createExclusionsPage({
	label = "Non-Edible Ingredients",
	description = "Ingredients that can't be directly consumed via equipping.",
	leftListLabel = "Non-Edible Ingredients",
	rightListLabel = "Ingredients",
	variable = mwse.mcm.createTableVariable{
		id = "nonEdible",
		table = config,
	},
	-- This filters the right list by ingredient object
	filters = {
		{ 
			label = "Ingredients", 
			type = "Object", 
			objectType = tes3.objectType.ingredient 
		},
	},
})
```

There is also a more customized way to filter the right list. The following code filters the right list by ingredients with value higher than 100 gold.

```lua linenums="1" title="My Awesome Mod/main.lua"
	filters = {
		{
			label = "Ingredients",
			callback = function()
				local ingreds = {}
				--- @param ingred tes3ingredient
				for ingred in tes3.iterateObjects(tes3.objectType.ingredients) do
					if ingred.value >= 100 then
						table.insert(ingreds, ingred)
					end
				end
				table.sort(ingreds)
				return ingreds
			end
		},
	},
```

docs/source/types/tes3aiConfig.md
# tes3aiConfig
<div class="search_terms" style="display: none">tes3aiconfig, aiconfig</div>

<!---
	This file is autogenerated. Do not edit this file manually. Your changes will be ignored.
	More information: https://github.com/MWSE/MWSE/tree/master/docs
-->

A game object which contains AI configuration information for an actor. This includes: alarm, fight, flee, hello values, which services the actor provides, goods types the actor trades in, and possible destinations this actor can take the player to.

## Properties

### `alarm`
<div class="search_terms" style="display: none">alarm</div>

Friendly access to actor's alarm value.

**Returns**:

* `result` (number)

***

### `bartersAlchemy`
<div class="search_terms" style="display: none">bartersalchemy</div>

Friendly access to a flag that controls whether the actor barters in potions.

**Returns**:

* `result` (boolean)

***

### `bartersApparatus`
<div class="search_terms" style="display: none">bartersapparatus</div>

Friendly access to a flag that controls whether the actor barters in alchemical apparati.

**Returns**:

* `result` (boolean)

***

### `bartersArmor`
<div class="search_terms" style="display: none">bartersarmor</div>

Friendly access to a flag that controls whether the actor barters in armor.

**Returns**:

* `result` (boolean)

***

### `bartersBooks`
<div class="search_terms" style="display: none">bartersbooks</div>

Friendly access to a flag that controls whether the actor barters in books.

**Returns**:

* `result` (boolean)

***

### `bartersClothing`
<div class="search_terms" style="display: none">bartersclothing</div>

Friendly access to a flag that controls whether the actor barters in clothing.

**Returns**:

* `result` (boolean)

***

### `bartersEnchantedItems`
<div class="search_terms" style="display: none">bartersenchanteditems</div>

Friendly access to a flag that controls whether the actor barters in enchanted items.

**Returns**:

* `result` (boolean)

***

### `bartersIngredients`
<div class="search_terms" style="display: none">bartersingredients</div>

Friendly access to a flag that controls whether the actor barters in alchemical ingredients.

**Returns**:

* `result` (boolean)

***

### `bartersLights`
<div class="search_terms" style="display: none">barterslights</div>

Friendly access to a flag that controls whether the actor barters in lights.

**Returns**:

* `result` (boolean)

***

### `bartersLockpicks`
<div class="search_terms" style="display: none">barterslockpicks</div>

Friendly access to a flag that controls whether the actor barters in lockpicks.

**Returns**:

* `result` (boolean)

***

### `bartersMiscItems`
<div class="search_terms" style="display: none">bartersmiscitems</div>

Friendly access to a flag that controls whether the actor barters in miscellaneous items.

**Returns**:

* `result` (boolean)

***

### `bartersProbes`
<div class="search_terms" style="display: none">bartersprobes</div>

Friendly access to a flag that controls whether the actor barters in probes.

**Returns**:

* `result` (boolean)

***

### `bartersRepairTools`
<div class="search_terms" style="display: none">bartersrepairtools</div>

Friendly access to a flag that controls whether the actor barters in repair hammers and prongs.

**Returns**:

* `result` (boolean)

***

### `bartersWeapons`
<div class="search_terms" style="display: none">bartersweapons</div>

Friendly access to a flag that controls whether the actor barters in weapons.

**Returns**:

* `result` (boolean)

***

### `fight`
<div class="search_terms" style="display: none">fight</div>

Friendly access to actor's fight value.

**Returns**:

* `result` (number)

***

### `flee`
<div class="search_terms" style="display: none">flee</div>

Friendly access to actor's flee value.

**Returns**:

* `result` (number)

***

### `hello`
<div class="search_terms" style="display: none">hello</div>

Friendly access to actor's hello value.

**Returns**:

* `result` (number)

***

### `merchantFlags`
<div class="search_terms" style="display: none">merchantflags</div>

Numeric access to all merchant flags. Shouldn't be used directly.

**Returns**:

* `result` (number)

***

### `offersEnchanting`
<div class="search_terms" style="display: none">offersenchanting</div>

Friendly access to a flag that controls whether the actor offers enchanting services.

**Returns**:

* `result` (boolean)

***

### `offersRepairs`
<div class="search_terms" style="display: none">offersrepairs</div>

Friendly access to a flag that controls whether the actor offers item repairs service.

**Returns**:

* `result` (boolean)

***

### `offersSpellmaking`
<div class="search_terms" style="display: none">offersspellmaking</div>

Friendly access to a flag that controls whether the actor offers spellmaking service.

**Returns**:

* `result` (boolean)

***

### `offersSpells`
<div class="search_terms" style="display: none">offersspells</div>

Friendly access to a flag that controls whether the actor sells spells.

**Returns**:

* `result` (boolean)

***

### `offersTraining`
<div class="search_terms" style="display: none">offerstraining</div>

Friendly access to a flag that controls whether the actor offers training services.

**Returns**:

* `result` (boolean)

***

### `travelDestinations`
<div class="search_terms" style="display: none">traveldestinations</div>

A table access to all [`tes3travelDestinationNode`](https://mwse.github.io/MWSE/types/tes3travelDestinationNode/) objects this actor can take the player to.

**Returns**:

* `result` ([tes3travelDestinationNode](../types/tes3travelDestinationNode.md)[])



d3d8/configuration.h
#pragma once

#define MASK(x) (1 << x)

// MGE Generic flags
#define MGE_DISABLED_BIT        0
#define MGE_DISABLED            MASK(MGE_DISABLED_BIT)
#define FOG_ENABLED_BIT         1
#define FOG_ENABLED             MASK(FOG_ENABLED_BIT)
#define FPS_COUNTER_BIT         2
#define FPS_COUNTER             MASK(FPS_COUNTER_BIT)
#define DISPLAY_MESSAGES_BIT    3
#define DISPLAY_MESSAGES        MASK(DISPLAY_MESSAGES_BIT)
#define USE_HW_SHADER_BIT       4
#define USE_HW_SHADER           MASK(USE_HW_SHADER_BIT)
#define NO_MW_SUNGLARE_BIT      5
#define NO_MW_SUNGLARE          MASK(NO_MW_SUNGLARE_BIT)
#define INPUT_LAG_FIX_BIT       6
#define INPUT_LAG_FIX           MASK(INPUT_LAG_FIX_BIT)
#define USE_MENU_CACHING_BIT    7
#define USE_MENU_CACHING        MASK(USE_MENU_CACHING_BIT)
#define ZOOM_ASPECT_BIT         8
#define ZOOM_ASPECT             MASK(ZOOM_ASPECT_BIT)
#define MWSE_DISABLED_BIT       9
#define MWSE_DISABLED           MASK(MWSE_DISABLED_BIT)
#define USE_FFESHADER_BIT       10
#define USE_FFESHADER           MASK(USE_FFESHADER_BIT)
#define TRANSPARENCY_AA_BIT     11
#define TRANSPARENCY_AA         MASK(TRANSPARENCY_AA_BIT)
#define USE_HDR_BIT             12
#define USE_HDR                 MASK(USE_HDR_BIT)
#define FPS_HOLD_BIT            13
#define FPS_HOLD                MASK(FPS_HOLD_BIT)
#define SKIP_INTRO_BIT          14
#define SKIP_INTRO              MASK(SKIP_INTRO_BIT)
#define CPU_IDLE_BIT            15
#define CPU_IDLE                MASK(CPU_IDLE_BIT)
#define SHADER_DEPTH_BIT        16
#define SHADER_DEPTH            MASK(SHADER_DEPTH_BIT)
// Distant Land flags
#define USE_DISTANT_LAND_BIT    17
#define USE_DISTANT_LAND        MASK(USE_DISTANT_LAND_BIT)
#define USE_DISTANT_STATICS_BIT 18
#define USE_DISTANT_STATICS     MASK(USE_DISTANT_STATICS_BIT)
#define NO_INTERIOR_DL_BIT      19
#define NO_INTERIOR_DL          MASK(NO_INTERIOR_DL_BIT)
#define REFLECTIVE_WATER_BIT    20
#define REFLECTIVE_WATER        MASK(REFLECTIVE_WATER_BIT)
#define REFLECT_NEAR_BIT        21
#define REFLECT_NEAR            MASK(REFLECT_NEAR_BIT)
#define REFLECT_INTERIOR_BIT    22
#define REFLECT_INTERIOR        MASK(REFLECT_INTERIOR_BIT)
#define NOT_USING_DL_BIT        23
#define NOT_USING_DL            MASK(NOT_USING_DL_BIT)
#define NO_MW_MGE_BLEND_BIT     24
#define NO_MW_MGE_BLEND         MASK(NO_MW_MGE_BLEND_BIT)
#define REFLECT_SKY_BIT         25
#define REFLECT_SKY             MASK(REFLECT_SKY_BIT)
#define DYNAMIC_RIPPLES_BIT     26
#define DYNAMIC_RIPPLES         MASK(DYNAMIC_RIPPLES_BIT)
#define BLUR_REFLECTIONS_BIT    27
#define BLUR_REFLECTIONS        MASK(BLUR_REFLECTIONS_BIT)
#define EXP_FOG_BIT             28
#define EXP_FOG                 MASK(EXP_FOG_BIT)

#define USE_ATM_SCATTER_BIT     29
#define USE_ATM_SCATTER         MASK(USE_ATM_SCATTER_BIT)
#define USE_GRASS_BIT           30
#define USE_GRASS               MASK(USE_GRASS_BIT)
#define USE_SHADOWS_BIT         31
#define USE_SHADOWS             MASK(USE_SHADOWS_BIT)



typedef unsigned long DWORD;
typedef unsigned char BYTE;
#pragma once

struct ConfigurationStruct {
    DWORD MGEFlags;
    BYTE AALevel;
    BYTE ZBufFormat;
    BYTE VWait;
    BYTE RefreshRate;
    bool Borderless;
    BYTE AnisoLevel;
    BYTE ScaleFilter;
    float LODBias;
    float ScreenFOV;
    BYTE FogMode;
    BYTE SSFormat;
    BYTE SSSuffix;
    char SSDir[208];
    char SSName[32];
    float HDRReactionSpeed;
    DWORD PerPixelLightFlags;
    int StatusTimeout;
    bool Force3rdPerson;
    struct {
        float x, y, z;
    } Offset3rdPerson;
    bool CrosshairAutohide;
    float UIScale;

    struct {
        float zoom, zoomRate, zoomRateTarget;
        bool rotateUpdate;
        float rotation, rotationRate;
        bool shake;
        float shakeMagnitude, shakeAccel;
    } CameraEffects;

    struct {
        float DrawDist;
        float NearStaticEnd;
        float FarStaticEnd;
        float VeryFarStaticEnd;
        float FarStaticMinSize;
        float VeryFarStaticMinSize;
        float AboveWaterFogStart;
        float AboveWaterFogEnd;
        float BelowWaterFogStart;
        float BelowWaterFogEnd;
        float InteriorFogStart;
        float InteriorFogEnd;
        float ExpFogDistMult;
        BYTE WaterWaveHeight;
        BYTE WaterCaustics;
        BYTE WaterReflect;
        DWORD ShadowResolution;
        float Wind[10];
        float FogD[10];
        float FgOD[10];
    } DL;

    struct {
        float SunMult[10];
        float AmbMult[10];
    } Lighting;

    struct {
        bool AltCombat;
        char Macros[4096];
        char Triggers[4096];
        char Remap[4096];
    } Input;

    char ShaderChain[512];

    bool LoadSettings();
};

extern ConfigurationStruct Configuration;
